python/pyarrow/gandiva.pyx

```diff
@@ -0,0 +1,184 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# cython: profile=False
+# distutils: language = c++
+# cython: embedsignature = True
+
+from libcpp cimport bool as c_bool, nullptr
+from libcpp.memory cimport shared_ptr, unique_ptr, make_shared
+from libcpp.string cimport string as c_string
+from libcpp.vector cimport vector as c_vector
+from libc.stdint cimport int64_t, uint8_t, uintptr_t
+
+from pyarrow.includes.libarrow cimport *
+from pyarrow.compat import frombytes
+from pyarrow.lib cimport (check_status, pyarrow_wrap_array)
+
+from pyarrow.includes.libgandiva cimport (CCondition, CExpression,
+                                          CNode, CProjector, CFilter,
+                                          CSelectionVector,
+                                          TreeExprBuilder_MakeExpression,
+                                          TreeExprBuilder_MakeFunction,
+                                          TreeExprBuilder_MakeLiteral,
+                                          TreeExprBuilder_MakeField,
+                                          TreeExprBuilder_MakeIf,
+                                          TreeExprBuilder_MakeCondition,
+                                          SelectionVector_MakeInt32,
+                                          Projector_Make,
+                                          Filter_Make)
+
+from pyarrow.lib cimport (Array, DataType, Field, MemoryPool,
+                          RecordBatch, Schema)
+
+cdef class Node:
+    cdef:
+        shared_ptr[CNode] node
+
+    cdef void init(self, shared_ptr[CNode] node):
+        self.node = node
+
+cdef make_node(shared_ptr[CNode] node):
+    cdef Node result = Node()
+    result.init(node)
+    return result
+
+cdef class Expression:
+    cdef:
+        shared_ptr[CExpression] expression
+
+    cdef void init(self, shared_ptr[CExpression] expression):
+        self.expression = expression
+
+cdef class Condition:
+    cdef:
+        shared_ptr[CCondition] condition
+
+    cdef void init(self, shared_ptr[CCondition] condition):
+        self.condition = condition
+
+cdef make_condition(shared_ptr[CCondition] condition):
```

Comment:
Could You use staticmethod factories with cython's fast [instantiation](), in similar fashion like [this](https://github.com/apache/arrow/pull/2873/files#diff-283f88a2b668d61530832cd5a2443377R1152)? We can call it either `create` or `from_ptr`.

Reply:
We also need to prevent instantiating python objects with uninitialized cdef properties. It requires to use both cython's fast instantiation and raising an exception from `__init__` like [here](https://github.com/apache/arrow/blob/master/python/pyarrow/array.pxi#L342).

Reply:
Thanks, I fixed that!